File: cpp/avoid_comments.cpp
// Avoid Comments
// Prefer to avoid comments rather than to write them
// If a comment is truly necessary, explain the "why" not the "what".

#include <iostream>
#include <cmath>

// Before
class CalculatorBefore {
public:
    double calculateArea(double radius) {
        // Calculate the area of a circle
        double area = M_PI * radius * radius;

        // Round the area to two decimal places
        area = std::round(area * 100.0) / 100.0;

        return area;
    }
};

// After
class Calculator {
public:
    double calculateArea(double radius) {
        double area = circleArea(radius);
        return std::round(area * 100.0) / 100.0;
    }

private:
    double circleArea(double radius) {
        return M_PI * radius * radius;
    }
};

int main() {
    CalculatorBefore calculatorBefore;
    std::cout << calculatorBefore.calculateArea(5) << std::endl; // Outputs: 78.54

    std::cout << std::endl;

    Calculator calculator;
    std::cout << calculator.calculateArea(5) << std::endl; // Outputs: 78.54

    return 0;
}

File: rust/avoid_comments.rs
// Avoid Comments
// Prefer to avoid comments rather than to write them
// If a comment is truly necessary, explain the "why" not the "what".

use std::f64::consts::PI;

// Before
struct CalculatorBefore;

impl CalculatorBefore {
    fn calculate_area(&self, radius: f64) -> f64 {
        // Calculate the area of a circle
        let area = PI * radius * radius;

        // Round the area to two decimal places
        let area = (area * 100.0).round() / 100.0;

        area
    }
}

// After
struct Calculator;

impl Calculator {
    fn calculate_area(&self, radius: f64) -> f64 {
        let area = self.circle_area(radius);
        (area * 100.0).round() / 100.0
    }

    fn circle_area(&self, radius: f64) -> f64 {
        PI * radius * radius
    }
}

fn main() {
    let calculator_before = CalculatorBefore;
    println!("{}", calculator_before.calculate_area(5.0)); // Outputs: 78.54

    println!();

    let calculator = Calculator;
    println!("{}", calculator.calculate_area(5.0)); // Outputs: 78.54
}
